File: mobile/test/auth_session_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:restaurant_mobile/core/auth/auth_session.dart';

AuthSession phien(
        {required DateTime hetHan, String token = 'jwt.rat.bi.mat'}) =>
    AuthSession(
      accessToken: token,
      expiresAt: hetHan,
      user: const AuthUser(
        userId: 'u1',
        fullName: 'Nguyễn Văn A',
        email: 'a@example.com',
        role: 'Customer',
      ),
    );

void main() {
  final moc = DateTime.utc(2026, 8, 20, 12, 0, 0);

  group('AuthSession.conHieuLuc', () {
    test('token còn hạn dài thì hợp lệ', () {
      expect(phien(hetHan: moc.add(const Duration(hours: 2))).conHieuLuc(moc),
          isTrue);
    });

    test('token đã quá hạn thì không hợp lệ', () {
      expect(
          phien(hetHan: moc.subtract(const Duration(minutes: 1)))
              .conHieuLuc(moc),
          isFalse);
    });

    test('token SẮP hết hạn (20 giây nữa) bị coi là hết hạn', () {
      // Đây là lý do tồn tại của biên an toàn. Bỏ biên đi thì ca này xanh trở lại nhưng người
      // dùng nhận 401 giữa lúc gửi đơn: request bay đi lúc token còn sống, tới nơi thì đã chết.
      expect(
          phien(hetHan: moc.add(const Duration(seconds: 20))).conHieuLuc(moc),
          isFalse);
    });

    test('so sánh theo UTC, không theo giờ máy', () {
      // Máy đặt sai múi giờ là chuyện có thật. Nếu so theo giờ địa phương, cùng một token sẽ
      // "còn hạn" ở múi này và "hết hạn" ở múi kia.
      final hetHanGioDiaPhuong = moc.add(const Duration(hours: 2)).toLocal();
      expect(phien(hetHan: hetHanGioDiaPhuong).conHieuLuc(moc), isTrue);
    });
  });

  group('không lộ token', () {
    test('toString() KHÔNG chứa access token', () {
      final s = phien(hetHan: moc, token: 'CHUOI_BI_MAT_KHONG_DUOC_IN');
      expect(s.toString(), isNot(contains('CHUOI_BI_MAT_KHONG_DUOC_IN')));
    });

    test('toString() vẫn đủ dùng để gỡ lỗi', () {
      // Nếu không kiểm điều này thì cách "an toàn" nhất là trả chuỗi rỗng, và người sau sẽ thêm
      // token vào cho dễ debug. Giữ nó vừa an toàn vừa có ích thì mới không bị sửa ngược.
      final s = phien(hetHan: moc);
      expect(s.toString(), contains('a@example.com'));
      expect(s.toString(), contains('Customer'));
    });
  });

  group('JSON', () {
    test('đi và về giữ nguyên token, hạn (UTC) và người dùng', () {
      final goc = phien(hetHan: moc.add(const Duration(hours: 3)));
      final ve = AuthSession.fromJson(goc.toJson());
      expect(ve.accessToken, goc.accessToken);
      expect(ve.expiresAt.isUtc, isTrue);
      expect(ve.expiresAt, goc.expiresAt);
      expect(ve.user.email, 'a@example.com');
      expect(ve.user.role, 'Customer');
    });

    test('chịu được Instant 9 chữ số thập phân — dạng backend THẬT trả về', () {
      // Đo từ backend Java đang chạy, không phải đoán:
      //
      //     "expiresAt":"2026-08-20T15:24:15.752877577Z"
      //
      // Instant của Java in tới nanosecond; DateTime của Dart chỉ tới microsecond nên cắt bớt 3
      // chữ số cuối. Ca này chốt rằng việc cắt đó không ném lỗi và không lệch giây.
      //
      // Bản đầu tôi viết test bằng chuỗi gọn tự nghĩ ra ("...:05Z"), tức đang kiểm app với dữ
      // liệu do chính app tưởng tượng chứ không phải dữ liệu máy chủ thật gửi.
      final ve = AuthSession.fromJson(const {
        'accessToken': 'abc',
        'expiresAt': '2026-08-20T15:24:15.752877577Z',
        'user': {
          'userId': 'u1',
          'fullName': 'Nguyễn Văn A',
          'email': 'a@example.com',
          'role': 'Customer',
        },
      });
      expect(ve.expiresAt.isUtc, isTrue);
      expect(ve.expiresAt, DateTime.utc(2026, 8, 20, 15, 24, 15, 752, 877));
    });
  });
}

File: mobile-rn/src/core/auth/__tests__/authSession.test.ts
import { BIEN_AN_TOAN_MS, authSessionTuJson, conHieuLuc, moTaSession } from '../authSession';

const NGUOI = { userId: 'u1', fullName: 'Khách', email: 'k@local.test', role: 'Customer' };

function phien(expiresAt: string) {
  return { accessToken: 'tok.abc.123', expiresAt, user: NGUOI };
}

describe('còn hiệu lực', () => {
  const bayGio = new Date('2026-08-23T10:00:00Z');

  it('token còn xa hạn thì dùng được', () => {
    expect(conHieuLuc(phien('2026-08-23T11:00:00Z'), bayGio)).toBe(true);
  });

  it('token đã quá hạn thì không', () => {
    expect(conHieuLuc(phien('2026-08-23T09:59:00Z'), bayGio)).toBe(false);
  });

  it('token còn 30 giây bị coi như đã hết hạn', () => {
    // Request bay đi, mạng 3G trong quán mất 2–3 giây, tới nơi thì token đã chết và người dùng
    // nhận 401 giữa lúc đang đặt món.
    expect(conHieuLuc(phien('2026-08-23T10:00:30Z'), bayGio)).toBe(false);
  });

  it('đúng mép biên an toàn thì chưa dùng được', () => {
    const mep = new Date(bayGio.getTime() + BIEN_AN_TOAN_MS).toISOString();
    expect(conHieuLuc(phien(mep), bayGio)).toBe(false);
  });

  it('hạn không đọc được thì coi như hết hạn, không coi như còn', () => {
    // Fail closed. Đọc nhầm theo hướng "còn hạn" khiến app gửi token chết và nhận 401 ở một chỗ
    // ngẫu nhiên; đọc theo hướng "hết hạn" chỉ bắt đăng nhập lại.
    expect(conHieuLuc(phien('không-phải-ngày'), bayGio)).toBe(false);
  });
});

describe('đọc từ JSON', () => {
  it('giờ địa phương được quy về UTC', () => {
    // Backend trả Instant có hậu tố Z. Nếu giữ theo giờ máy thì một thiết bị đặt sai múi giờ sẽ
    // tự cho là token còn hạn hoặc đã hết hạn sớm vài tiếng.
    const s = authSessionTuJson({
      accessToken: 't',
      expiresAt: '2026-08-23T17:00:00+07:00',
      user: NGUOI,
    });
    expect(s.expiresAt).toBe('2026-08-23T10:00:00.000Z');
  });

  it('hạn hỏng thì NÉM, để nơi gọi xoá phiên thay vì giữ một phiên vô nghĩa', () => {
    expect(() => authSessionTuJson({ accessToken: 't', expiresAt: 'rác', user: NGUOI })).toThrow();
  });
});

describe('mô tả phiên', () => {
  it('KHÔNG chứa token', () => {
    // Khác với Dart, mặc định của JavaScript là in HẾT mọi trường, nên token lộ ngay từ lần
    // console.log đầu tiên. Hàm này tồn tại để không ai phải log nguyên object.
    const mo = moTaSession(phien('2026-08-23T11:00:00Z'));
    expect(mo).not.toContain('tok.abc.123');
    expect(mo).toContain('k@local.test');
  });
});
